python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -331,3 +331,81 @@ def get_units(value):
         return _unit_map[value.lower()]
 
     return value
+
+
+class DateOffset(object):
+    def __init__(self, n=1, normalize=False, **kwds):
+
+        self.resolutions = [
+            "days",
+            "hours",
+            "microseconds",
+            "minutes",
+            "months",
+            "nanoseconds",
+            "seconds",
+            "weeks",
+            "years",
+        ]
+
+        self.n = n
+
+        if normalize:
+            raise NotImplementedError("normalize not supported for DateOffset")
+
+        # TODO: Pandas supports combinations
+        if len(kwds) > 1:
+            raise ValueError("only a single unit may" "be specified at a time")
+
+        # in pandas, **kwds is passed to an underling relativedelta
+        # which has its own valid kwargs. so must manually validate
+        supported_kwargs = ["months"]
+        wrong_kwargs = set(kwds.keys()).difference(supported_kwargs)
+        if len(wrong_kwargs) > 0:
+            raise ValueError(
+                f"Keyword arguments '{','.join(list(wrong_kwargs))}'"
+                " are not supported"
+            )
+
+        # something like years=1.5 was passed
+        if kwds.get("years") and not kwds.get("years") == int(
+            kwds.get("years")
+        ):
+            raise ValueError("fractional years not supported")
+
+        # something like months=1.5 was passed
+        if kwds.get("months") and not kwds.get("months") == int(
+            kwds.get("months")
+        ):
+            raise ValueError("fractional months not supported")
+
+        for r in self.resolutions:
+            setattr(self, r, kwds.get(r, None))
+
+        # some logic could be implemented here for more complex cases
+        # such as +1 year, -12 months
+        self._is_no_op = False
+        if all([getattr(self, r) in (None, 0) for r in self.resolutions]):
+            self._is_no_op = True
+
+    def _generate_column(self, size, op):
+        months = -self.months if op == "sub" else self.months
+        col = cudf.core.column.as_column(
+            months, dtype=np.dtype("int16"), length=size
+        )
+        return col
```

Comment:
Does libcudf have an overload for the API that takes in a device scalar instead of a column? If not could you raise an issue?

Reply:
raised https://github.com/rapidsai/cudf/issues/6990